src/analysis/mesh/qgsmeshtriangulation.cpp

```diff
@@ -0,0 +1,316 @@
+/***************************************************************************
+                          qgsmeshtriangulation.cpp
+                          -----------------
+    begin                : August 9th, 2020
+    copyright            : (C) 2020 by Vincent Cloarec
+    email                : vcloarec at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmeshtriangulation.h"
+#include "qgsdualedgetriangulation.h"
+#include "qgsvectorlayer.h"
+#include "qgscoordinatetransformcontext.h"
+#include "qgscurve.h"
+#include "qgscurvepolygon.h"
+#include "qgsmultisurface.h"
+#include "qgsmulticurve.h"
+#include "qgsfeedback.h"
+
+QgsMeshTriangulation::QgsMeshTriangulation(): QObject()
+{
+  mTriangulation.reset( new QgsDualEdgeTriangulation() );
+}
+
+
+QgsMeshTriangulation::~QgsMeshTriangulation() = default;
+
+
+bool QgsMeshTriangulation::addVertices( QgsVectorLayer *vectorLayer,
```

Comment:
It would be better to accept a QgsFeatureSource or QgsFeatureIterator here instead of a vector layer. That would keep the API more flexible and thread safe (accessing a QgsVectorLayer from the project in a background thread isn't safe, but if you get the feature source or iterator in the main thread first then it's safe to pass to a background thread)